### DISCUSSION TITLE
operating hours counter (Betriebsstundenzähler)

### DISCUSSION THREAD

**Discussion by MichiM2** [arendst/Tasmota#17783] [Ideas]
Hi,
sometimes i miss a feature in TASMOTA, which measures elapsed time. 
a) It could be defined as type of an GPIO, similar to the Counter-feature and will measure the elapsed time, die GPIO is ON. 
b) Or/and  it could be defined somehow to measure the duration of a POWER-Output state ON.

The MQTT-Output could be similar to the ENERGY-Output:

ENERGY":{"TotalStartTime":"2020-12-26T12:04:31","Total":407.012,"Yesterday":10.465,"Today":5.376}

It could be used e.g. for measuring the consumption of an Oil-Burner, which should proportional to the ON-Time. 

Michi

**Comment by barbudor** [comment]
You can enable a dummy energy device that will count a fixed power consumption when the relay is on
So if your gpio is a switch it will activate a dummy relay and the energy meter will count
If you set the power consumption to 1 W then energy will be the number of hours
If you set it to 60, it will be minutes..

I can't remember right now how to activate that but I can look later

**Comment by MichiM2** [comment]
That would be an elegant way. But i have no idea how to enable a "dummy energy device" and set it to a fix consumption.

So it would be nice, if you could tell me how to do this.

**Reply by barbudor** [reply]
Yes, I thought it would be very cruel to tell you that was possible but not giving the recipe :rofl: 
But I was at work, answering from my phone and actually this is not documented

To activate the dummy energy monitoring driver, you need to set an unused GPIO to "Option_A 2" (Option_A then index 2)
Then use the commands `VoltageSet` (V), `CurrentSet` (mA), `PowerSet` (W) to simulate what you actual load is using and Tasmota will count the energy when the relay is ON as if it was measuring it
Exemple:
`backlog VoltageSet 240; CurrentSet 5000; PowerSet 1200` simulate a fully resistive load at 1200W
`backlog VoltageSet 240; CurrentSet 5000; PowerSet 800` simulate a load wiht a 0.67 power factor

Enjoy

**Reply by BastiJoe** [reply]
That's actually a nice feature. I am setting this up for an oil burner as well. @barbudor Thank you for your support.
I tried with Option A index 2 put it doesn't work as expected:

13:26:58.981 CMD: backlog VoltageSet 1; CurrentSet 1; PowerSet 1
13:26:58.984 SRC: WebConsole from 192.168.2.134
13:26:58.987 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 38, Pld -99, Data 'VoltageSet 1; CurrentSet 1; PowerSet 1'
13:26:59.056 SRC: Backlog
13:26:59.059 CMD: Grp 0, Cmd 'VOLTAGESET', Idx 1, Len 1, Pld 1, Data '1'
13:26:59.064 RSL: RESULT = {"Command":"Unknown"}
13:26:59.271 SRC: Backlog
13:26:59.274 CMD: Grp 0, Cmd 'CURRENTSET', Idx 1, Len 1, Pld 1, Data '1'
13:26:59.279 RSL: RESULT = {"Command":"Unknown"}
13:26:59.566 SRC: Backlog
13:26:59.569 CMD: Grp 0, Cmd 'POWERSET', Idx 1, Len 1, Pld 1, Data '1'
13:26:59.575 RSL: RESULT = {"Command":"Unknown"}
13:27:02.402 WIF: Sending Gratuitous ARP
13:27:04.591 WIF: Checking connection...

**Reply by barbudor** [reply]
Where did you applied the Option A2 ?
- In Configure -> Configure Module ? => That should be straight forward
- In Configure -> Configure Template ? => If so, you must insure the template is activated by checking the "Activate" check-box in Configure Other

As soon as you do that, you should have the usual Energy view on the main page and then you can use the commands as described

**Reply by BastiJoe** [reply]
I configured in In Configure -> Configure Module
![image](https://user-images.githubusercontent.com/47536246/214570351-a8d5a81b-d015-4611-a219-16f074e6ecdf.png)
but i also do not have the enrgy view:
![image](https://user-images.githubusercontent.com/47536246/214570506-977b42d1-6290-402c-8e91-2ff648cf7f88.png)

**Comment by MichiM2** [comment]
i activated it yesterday and it did work immediately. But i think it is implemented in Tasmota only from V 9.3.1.2 on.

You have also to set a GPIO-Pin for Relay <x> where x corresponds to the <x> from the Option A <x>.

But i now have another problem here:
I can set 1V, 1A , 1W  with the command

backlog VoltageSet 1; CurrentSet 1000; PowerSet 1    ( Current is in mA !)

but i can not set it to 100V, 10A, 1000W.

Are there any preconditions or Limitations on these setting?






Am 25.01.23 um 13:48 schrieb Barbudor:

**Reply by BastiJoe** [reply]
Thanks.
ok, with the relay i get the energy monitor. But i acutally have an INPUT Pin, where i connect the oil burner with an optocoupler.
When i change the GPIO to Button or Switch instead of relay, the energy monitor disappears.

**Reply by BastiJoe** [reply]
can you set to 100V, 10000mA, 1000W ?

**Reply by barbudor** [reply]
Yes having a RELAY is a requirement
So you must have a GPIO defined as Relay even if not using
You can set you switch input in Follow mode (SwitchMode 1) or Invert Follow mode (SwitchMode 2) and then the pseudo relay will follow the switch input

**Comment by MichiM2** [comment]
The relay could be virtual. That means, you can use any UNUSED GPIO for that and there need not a real Relay to be connected.

With the INPUT-Pin you should control the virtual Relay.



Am 25.01.23 um 14:15 schrieb BastiJoe:

**Reply by BastiJoe** [reply]
This is working! Great, Thank you!
BTW i am going to measure the oil tank level wih an HC SR04.

**Reply by Noschvie** [reply]
> This is working! Great, Thank you!

please share screenshots of your working configuration, thanks!

**Reply by BastiJoe** [reply]
![image](https://user-images.githubusercontent.com/47536246/214607424-abbafe35-7296-4ff7-bc43-85820d2a29bc.png)
D5 is my INPUT Pin

**Reply by barbudor** [reply]
Tasmota is not supposed to support Relay not starting at 1
Why not using Relay 1 and Switch 1 ?

There is no relation between the 2 in "Option A2" and the relay index

**Reply by BastiJoe** [reply]
ok, i am still a bit confused with these indexes.
you are suggesting like this?
![image](https://user-images.githubusercontent.com/47536246/214612253-787fdd5a-446d-47b9-9627-c5c3c6081427.png)

**Comment by MichiM2** [comment]
no.

i can set it to 1V, 1000mA, 1W with

backlog VoltageSet 1; CurrentSet 1000; PowerSet 1

but when i try to set it to 1kW with

backlog VoltageSet 250; CurrentSet 4000; PowerSet 1000

the result is




or with

backlog VoltageSet 100; CurrentSet 10000; PowerSet 1000

the result is





Tasmota-Version is Tasmota 12.3.1 by Theo Arends <https://bit.ly/tasmota>



Am 25.01.23 um 14:22 schrieb BastiJoe:

**Comment by barbudor** [comment]
@MichiM2 There are multiple messages out or order on Github so I'm not sure which one is the latest and if you still have problems

**Reply by MichiM2** [reply]
Yes, i still have problems. The Method you told me initially (Dummy Energy Device)  works well.

I now have another problem here:
I can set 1V, 1A , 1W  with the command
backlog VoltageSet 1; CurrentSet 1000; PowerSet 1    ( Current is in mA !)

but i can not set it to 100V, 10A, 1000W.
Are there any preconditions or Limitations on these setting?

when i try to set it to 1000W with
backlog VoltageSet 250; CurrentSet 4000; PowerSet 1000

the result is

![grafik](https://user-images.githubusercontent.com/37748504/214612019-0af5d2bb-b033-41e1-a97e-19fe5a421dcd.png)

or with

backlog VoltageSet 100; CurrentSet 10000; PowerSet 1000

the result is

![grafik](https://user-images.githubusercontent.com/37748504/214612132-86ca3a92-618a-407c-b868-b3861fb43186.png)


PS: I am sorry, that the messages got out of order. I did erroneously answer the emails instead of answering her.

**Reply by barbudor** [reply]
> but i can not set it to 100V, 10A, 1000W.

I actually copied the line from your own message and pasted it in my console:
```
21:46:11.250 CMD: backlog VoltageSet 100; CurrentSet 10000; PowerSet 1000
21:46:11.282 MQT: stat/nodemcu/VOLTAGESET = {"VoltageSetCal":10000}
21:46:11.501 MQT: stat/nodemcu/CURRENTSET = {"CurrentSetCal":1000000}
21:46:11.703 MQT: stat/nodemcu/POWERSET = {"PowerSetCal":100000}
``` 
![image](https://user-images.githubusercontent.com/2996042/214686293-8b6320ea-733d-4182-8524-ac56d70d479c.png)

```
21:47:42.525 CMD: backlog VoltageSet 250; CurrentSet 4000; PowerSet 1000
21:47:42.544 MQT: stat/nodemcu/VOLTAGESET = {"VoltageSetCal":25000}
21:47:42.775 MQT: stat/nodemcu/CURRENTSET = {"CurrentSetCal":400000}
21:47:43.024 MQT: stat/nodemcu/POWERSET = {"PowerSetCal":100000}
```
![image](https://user-images.githubusercontent.com/2996042/214687365-17528ee9-df69-4e8b-a3d6-aefb72e1b7d8.png)

However
```
21:51:02.492 CMD: backlog VoltageSet 1000; CurrentSet 1000; PowerSet 1000
21:51:02.571 MQT: stat/nodemcu/VOLTAGESET = {"VoltageSetCal":1000}
21:51:02.783 MQT: stat/nodemcu/CURRENTSET = {"CurrentSetCal":100000}
21:51:03.040 MQT: stat/nodemcu/POWERSET = {"PowerSetCal":100000}
```
break the energy monitoring:
![image](https://user-images.githubusercontent.com/2996042/214687883-bc6dd3a0-f77f-44d4-a913-1a1d3930c80c.png)

But re-entiring one of the previous values does recover correctly

Nevertheless, you may have broken some calculation with trials that are for some reason non recoverable
If you can't recover a working behavior you may just have to erase the settings and reconfigure from scratch ([`reset 6`](https://tasmota.github.io/docs/Commands/#reset) will keep your Wifi and MQTT settings and erase the rest)

**Reply by BastiJoe** [reply]
I actually had a similar behavior as @MichiM2
The apparent power was always correct but the power factor and active power were adjusted to lower values. 
I will upload more details tomorrow.

**Reply by BastiJoe** [reply]
I tried your steps after a reset 6 and it is perfectly working.

**Reply by MichiM2** [reply]
after doing a reset 6 i tried again with 100V, 10A , 1000W , but Tasmota did not accept it:

```
10:02:15.060 CMD: backlog VoltageSet 100; CurrentSet 10000; PowerSet 1000
10:02:15.079 MQT: tasmota03/stat/RESULT = {"VoltageSetCal":24000}
10:02:15.310 MQT: tasmota03/stat/RESULT = {"CurrentSetCal":41666}
10:02:15.560 MQT: tasmota03/stat/RESULT = {"PowerSetCal":100000}

10:02:32.245 CMD: backlog VoltageSet2 100; CurrentSet2 10000; PowerSet2 1000
10:02:32.264 MQT: tasmota03/stat/RESULT = {"VoltageSetCal":24000}
10:02:32.497 MQT: tasmota03/stat/RESULT = {"CurrentSetCal":41666}
10:02:32.747 MQT: tasmota03/stat/RESULT = {"PowerSetCal":100000}
```
but 250V, 4A, 1000W did work:

```
10:03:23.707 CMD: backlog VoltageSet 250; CurrentSet 4000; PowerSet 1000
10:03:23.725 MQT: tasmota03/stat/RESULT = {"VoltageSetCal":25000}
10:03:23.956 MQT: tasmota03/stat/RESULT = {"CurrentSetCal":400000}
10:03:24.207 MQT: tasmota03/stat/RESULT = {"PowerSetCal":100000}

10:03:48.976 CMD: backlog VoltageSet2 250; CurrentSet2 4000; PowerSet2 1000
10:03:48.997 MQT: tasmota03/stat/RESULT = {"VoltageSetCal":25000}
10:03:49.228 MQT: tasmota03/stat/RESULT = {"CurrentSetCal":400000}
10:03:49.475 MQT: tasmota03/stat/RESULT = {"PowerSetCal":100000}
```

my Tasmota for this try was V 10.1.0

After all, i am happy with this solution. @barbudor and @BastiJoe Thanks for your support!

**Reply by barbudor** [reply]
All my tests were on latest dev 12.3.1.4

**Comment by jnissen** [comment]
Hello,
follow settings
` backlog VoltageSet 250; CurrentSet 4000; PowerSet 1000`

![2023-11-13 07_56_02-h001 - Main Menu – Mozilla Firefox](https://github.com/arendst/Tasmota/assets/7793064/604e83f0-2134-43c8-8cd4-500c2d6ff9b3)

`07:56:47.465 RSL: SENSOR = {"Time":"2023-11-13T07:56:47","Switch2":"ON","ENERGY":{"TotalStartTime":"2023-11-11T19:58:15","Total":35.997,"Yesterday":23.999,"Today":7.944,"Period":3,"Power":1000,"ApparentPower":1000,"ReactivePower":0,"Factor":1.00,"Frequency":50,"Voltage":250,"Current":4.000}}



07:56:57.471 RSL: SENSOR = {"Time":"2023-11-13T07:56:57","Switch2":"ON","ENERGY":{"TotalStartTime":"2023-11-11T19:58:15","Total":36.000,"Yesterday":23.999,"Today":7.947,"Period":3,"Power":1000,"ApparentPower":1000,"ReactivePower":0,"Factor":1.00,"Frequency":50,"Voltage":250,"Current":4.000}}

07:57:07.431 RSL: STATE = {"Time":"2023-11-13T07:57:07","Uptime":"0T00:43:00","UptimeSec":2580,"Heap":20,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"ON","Wifi":{"AP":1,"SSId":"sensors","BSSId":"00:1A:8C:2A:5C:C2","Channel":1,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:05"}}

07:57:07.441 RSL: SENSOR = {"Time":"2023-11-13T07:57:07","Switch2":"ON","ENERGY":{"TotalStartTime":"2023-11-11T19:58:15","Total":36.002,"Yesterday":23.999,"Today":7.949,"Period":3,"Power":1000,"ApparentPower":1000,"ReactivePower":0,"Factor":1.00,"Frequency":50,"Voltage":250,"Current":4.000}}


07:57:17.441 RSL: SENSOR = {"Time":"2023-11-13T07:57:17","Switch2":"ON","ENERGY":{"TotalStartTime":"2023-11-11T19:58:15","Total":36.005,"Yesterday":23.999,"Today":7.952,"Period":3,"Power":1000,"ApparentPower":1000,"ReactivePower":0,"Factor":1.00,"Frequency":50,"Voltage":250,"Current":4.000}}`


the log Message are sen dall 10 sec. 

How can I see the Opperating Hour ?
if I calculate the difference between Today":7.952," - Today":7.949," it was three 0.003 not 0.010

Where I my mistake ?

**Reply by MichiM2** [reply]
Hi,
i think your "mistake" is, that you calculate the difference between two values which indirectly represent hours. So the difference between 7.952 - 7.949 is 0.003 hours, which is (3600 Sec. *0.003) = 10.8 Seconds which is rather near the 10 Seconds-Interval Tasmota sends the two Telegrams. That could be because of roundings?

**Reply by jnissen** [reply]
Hello, 
Thanks for your reply. Do yout Know it is possible to add Decimalplaces more than 3 in Tasmota ?

Greetings from germany
Jörg
